src/Runtime/DynMemRef.h

```diff
@@ -29,14 +29,34 @@ typedef int64_t INDEX_TYPE;
 // all ranks and type combinations.
 // We will refer to it as a DMR (Dynamic MemRef).
 struct DynMemRef {
+
+  // Pointer to the raw memory space allocated to host the DMR content. This
+  // pointer should only be acessed for memory management purposes, not for
+  // reading DMR content.
```

Comment:
Such comments make me think that we would be better served by making this member private, possibly only under C++ flags, so that you can effectively control the data the way you want.

Love the commenting, very helpful.

Reply:
It occurs to me that there is a mismatch between visibility control (imposed by `public` or `private`) is somewhat different from the kind of restrictions I wish to impose as indicated in the comments. Ideally, here we want to limit access to certain data members considering the user's intent (e.g., don't use the `data` field for anything other than alloc/dealloc purposes). So this occurs to me as something better addressed by reasonable API design, which is a topic that hasn't yet been addressed, but is touched upon by the upcoming JNI patch.

I suggest that we revisit & address this issue when we outline our C++ API? Please let me know your thoughts.

Reply:
For the user facing API, at least for the JNI wrapper, I think there should be no `alignedData` and related functions. Since Java has no concept of alignment. Basically, for the Java `DynMemRef` API, the user will create the `DynMemRef` object, then call `set{Byte|Short|Int...}Data(array)` to set the data buffer, which internally sets the `data` field to the user supplied data array. Similarly on the returning path, it's assumed that the data buffer is at the `data` field in the native `DynMemRef` structure, which is used to construct a Java `DynMemRef` object.

In essence, alignment should be an internal matter. User input and output shouldn't be bothered with alignment, i.e., if there is alignment requirement for the computation, the native code should do the alignment internally once it gets the user input. I think the user facing C/C++ API probably should also take this route.

Reply:
Totally agree, we should definitely hide concerns such as data alignment when designing for high-level user-facing API.